### ISSUE TITLE
Question: How is nuitka dealing with python imports?

### ISSUE THREAD

**Issue by brupelo** [Nuitka/Nuitka#135]
@kayhayen Hello, If you recall I've already asked about this one on previous posts but the whole subject is still pretty confusing to me, so please allow me to come back to it once again.

I'd really like to understand how nuitka is dealing with python imports. While I'm fine considering other parts of nuitka code generation/optimization as a black box (at least for now) I'd really like to understand how this is working, I'd really like to understand fully the real meaning of the import control options group e.g. `{--recurse-stdlib --recurse-none --recurse-all, --recurse-on --recurse-to --recurse-not-to --recurse-plugins --recurse-files, --recurse-pattern}`. 

From what I've roughly understood so far, the fact than nuitka parses/discover more or less imports won't affect to the `PE Import Directory`, right? Said otherwise, whether nuitka add more or less c++ that imports python stuff is just affecting the PE `.text/.data` sections, is this correct? Or... I'm just talking nonsense, I'm really confused about the whole subject :/

Anyway, summing up, consider this a very low-priority question it'd be really interesting if eventually you feel like it and write about it. In the form of blog post, an explanation here, in the dev manual... Being able to understand this will allow me to think about possible addons to add on top of nuitka that may or may not be relevant but definitely will give me a wider perspective about the whole thing.

Regards,
B

**Comment by brupelo** [comment]
Brief note one about this one, it took me a little bit of time to fully understand the whole recursive options but after playing/fooling around and analizing the type of c code that was being generated out of a dummy hello world package things **I've fully understood** what they are doing.

Just adding this remark so you can decide whether is still important to eventually deal with this one or not. Maybe there are users more clever than me than by just reading the Options help will immediately grasp what they're intended for. In this particular case, I'm more a practical guy that learn by "doing" and checking what the code does ;)

\o

**Comment by kayhayen** [comment]
I sent an email to the mailing list about these, basically part of the --recurse-* options should be named more like --include, namely those that address filenames and directories, because that is what those do, and other others should be --follow or --nofollow to decide automatic following of imports if they are used or not.

Until I enhanced this, I am not going to describe the current state though. It's wild growth, and with the advent of --include options has become unnecessary.

**Comment by kayhayen** [comment]
So at long last, here is an answer, that I intend to put into user manual afterwards:

Inclusion vs Following vs. Defaults
-----------------------------------------------

The filename you gave on the command line is always included. 

As for inclusion options: Inclusion of modules is supposed to only include that module no matter what. And of packages, that package, all modules below its namespace, no matter what. No other options will limit that. Your best attempt currently at excluding stuff defined to be included is to delete its Python files before compiling.

Following only happens when an import is encountered, starting from any included or followed to module, so that can happen recursively. You can enable or disable it globally. For standalone mode it's enabled globally, for module mode, it's disabled globally. But you can still make decisions for specific namespaces. When you say --nofollow-import-to that is going to take precedence over --follow-import-to no matter what.

Currently, to compile a package, you also have to include it with that option. That disables you from making overrides to it. A future release might change things, so that packages compiled are automatically included, but obey --nofollow-import-to=*.tests to make it easy. Alternatively a --noinclude=*.tests option with a high precedence could also come, this is not clear yet.


Obviously I hate documenting what I consider interim states, but there have been several people who wondered this already, and it's a fair question.
